### PR TITLE
Branch refactor and registration

### PR DESCRIPTION
**Pull request by kumare3**
This currently fails during execution

### CODE REVIEW

File: tests/flytekit/unit/annotated/test_type_hints.py [hunk 1]
```diff
@@ -477,46 +477,49 @@ def t2(a: str) -> str:
...
-                return x, d
-
-        foo()
+    assert x == (7, "Not Hello!")
+
+    x = my_wf(a=2, b="hello ")
+    assert x == (4, "It is hello")
+
+
+# def test_wf1_branches_no_else():
```
Comment (wild-endeavor) [inline tests/flytekit/unit/annotated/test_type_hints.py:499]: should this be here?

File: tests/flytekit/unit/annotated/test_type_hints.py [hunk 2]
```diff
@@ -531,10 +534,18 @@ def t2(a: str) -> str:
...
+            .then(t2(a=b))
+            .elif_(x >= 5)
+            .then(t2(a=y))
+            .else_()
+            .fail("All Branches failed")
+        )
         return x, d
 
-    with pytest.raises(AssertionError):
+    with pytest.raises(ValueError):
```
Comment (wild-endeavor) [inline tests/flytekit/unit/annotated/test_type_hints.py:548]: does the failure value make it into the ValueError at all?  Just wondering.

File: flytekit/annotated/condition.py [hunk 3]
```diff
@@ -1,104 +1,332 @@
...
+        return self._name
+
+    @property
+    def condition(self):
+        return self._condition
+
+    def if_(self, expr: bool) -> Case:
+        ctx = FlyteContext.current_context()
+        if ctx.compilation_state:
+            # Set some compilation context
```
Comment (wild-endeavor) [inline flytekit/annotated/condition.py:326]: does this need to be filled in?

File: flytekit/annotated/node.py [hunk 4]
```diff
@@ -68,6 +79,7 @@ def get_registerable_entity(self) -> SdkNode:
                 metadata=self._metadata,
                 sdk_launch_plan=self._flyte_entity.get_registerable_entity(),
             )
+        # TODO: Add new annotated LaunchPlan when done
```
Comment (wild-endeavor) [inline flytekit/annotated/node.py:82]: can remove this

File: flytekit/annotated/node.py [hunk 5]
```diff
@@ -60,6 +63,14 @@ def get_registerable_entity(self) -> SdkNode:
                 metadata=self._metadata,
                 sdk_workflow=self._flyte_entity.get_registerable_entity(),
             )
+        elif isinstance(self._flyte_entity, _workflow_model.BranchNode):
+            self._sdk_node = SdkNode(
+                self._id,
+                upstream_nodes=sdk_nodes,
+                bindings=self._bindings,
+                metadata=self._metadata,
+                sdk_branch=self._flyte_entity,
```
Comment (wild-endeavor) [inline flytekit/annotated/node.py:72]: What's the type of `self._flyte_entity` here?

**Review by wild-endeavor — COMMENTED** [review]
tests are failing with an error that doesn't make sense, though I didn't really look into it.

File: flytekit/annotated/node.py (hunk 5 above)
Comment (kumare3) [reply flytekit/annotated/node.py:72]: thats the thing, right now its BranchNode. I want to discuss this and clean this up

File: tests/flytekit/unit/annotated/test_type_hints.py (hunk 2 above)
Comment (kumare3) [reply tests/flytekit/unit/annotated/test_type_hints.py:548]: ya it does, we could make it something else. This is what it is today

File: tests/flytekit/unit/annotated/test_type_hints.py (hunk 1 above)
Comment (kumare3) [reply tests/flytekit/unit/annotated/test_type_hints.py:499]: i changed it a little bit and it works now

File: flytekit/annotated/node.py (hunk 4 above)
Comment (kumare3) [reply flytekit/annotated/node.py:83]: done

File: tests/flytekit/unit/annotated/test_serialization.py [hunk 6]
```diff
@@ -39,3 +42,70 @@ def raw_container_wf(val1: int, val2: int) -> int:
...
+        return x, f
+
+    ctx = FlyteContext.current_context()
+    registration_settings = context_manager.RegistrationSettings(
+        project="project", domain="domain", version="version", image="image", env=None,
+    )
+    with ctx.current_context().new_registration_settings(registration_settings=registration_settings):
+        wf = my_wf.get_registerable_entity()
+        assert wf is not None
+        print(wf)
```
Comment (wild-endeavor) [inline tests/flytekit/unit/annotated/test_serialization.py:78]: should we assert something here?

File: tests/flytekit/unit/annotated/test_serialization.py [hunk 7]
```diff
@@ -39,3 +42,70 @@ def raw_container_wf(val1: int, val2: int) -> int:
...
+    assert my_wf(a=2) == "hello"
+
+    ctx = FlyteContext.current_context()
+    registration_settings = context_manager.RegistrationSettings(
+        project="project", domain="domain", version="version", image="image", env=None,
+    )
+    with ctx.current_context().new_registration_settings(registration_settings=registration_settings):
+        wf = my_wf.get_registerable_entity()
+        assert wf is not None
+        print(wf)
```
Comment (wild-endeavor) [inline tests/flytekit/unit/annotated/test_serialization.py:109]: and here

**Review by wild-endeavor — COMMENTED** [review]
turn prints to asserts then merge?

**Pull request merged**